// angular2/ts/src/core/testability/browser_testability.js
System.register(['angular2/src/core/testability/testability', 'angular2/src/core/facade/lang'], function(exports_1) {
    var testability_1, lang_1;
    var PublicTestability, BrowserGetTestability;
    return {
        setters:[
            function (testability_1_1) {
                testability_1 = testability_1_1;
            },
            function (lang_1_1) {
                lang_1 = lang_1_1;
            }],
        execute: function() {
            class PublicTestability {
                constructor(testability) {
                    this._testability = testability;
                }
                isStable() { return this._testability.isStable(); }
                whenStable(callback) { this._testability.whenStable(callback); }
                findBindings(using, provider, exactMatch) {
                    return this.findProviders(using, provider, exactMatch);
                }
                findProviders(using, provider, exactMatch) {
                    return this._testability.findBindings(using, provider, exactMatch);
                }
            }
            class BrowserGetTestability {
                static init() { testability_1.setTestabilityGetter(new BrowserGetTestability()); }
                addToWindow(registry) {
                    lang_1.global.getAngularTestability = function (elem, findInAncestors = true) {
                        var testability = registry.findTestabilityInTree(elem, findInAncestors);
                        if (testability == null) {
                            throw new Error('Could not find testability for element.');
                        }
                        return new PublicTestability(testability);
                    };
                    lang_1.global.getAllAngularTestabilities = function () {
                        var testabilities = registry.getAllTestabilities();
                        return testabilities.map((testability) => { return new PublicTestability(testability); });
                    };
                }
            }
            BrowserGetTestability = BrowserGetTestability;
        }
    }
});
