// angular2/ts/src/core/forms/directives/number_value_accessor.js
System.register(['angular2/src/core/metadata', 'angular2/src/core/di', './control_value_accessor', 'angular2/src/core/facade/lang', './shared'], function(exports_1) {
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var metadata_1, di_1, control_value_accessor_1, lang_1, shared_1;
    var NUMBER_VALUE_ACCESSOR, NumberValueAccessor;
    return {
        setters:[
            function (metadata_1_1) {
                metadata_1 = metadata_1_1;
            },
            function (di_1_1) {
                di_1 = di_1_1;
            },
            function (control_value_accessor_1_1) {
                control_value_accessor_1 = control_value_accessor_1_1;
            },
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (shared_1_1) {
                shared_1 = shared_1_1;
            }],
        execute: function() {
            const NUMBER_VALUE_ACCESSOR = lang_1.CONST_EXPR(new di_1.Provider(control_value_accessor_1.NG_VALUE_ACCESSOR, { useExisting: di_1.forwardRef(() => NumberValueAccessor), multi: true }));
            /**
             * The accessor for writing a number value and listening to changes that is used by the
             * {@link NgModel}, {@link NgFormControl}, and {@link NgControlName} directives.
             *
             *  # Example
             *  ```
             *  <input type="number" [(ng-model)]="age">
             *  ```
             */
            let NumberValueAccessor = class {
                constructor(_renderer, _elementRef) {
                    this._renderer = _renderer;
                    this._elementRef = _elementRef;
                    this.onChange = (_) => { };
                    this.onTouched = () => { };
                }
                writeValue(value) { shared_1.setProperty(this._renderer, this._elementRef, 'value', value); }
                registerOnChange(fn) {
                    this.onChange = (value) => { fn(lang_1.NumberWrapper.parseFloat(value)); };
                }
                registerOnTouched(fn) { this.onTouched = fn; }
            };
            NumberValueAccessor = __decorate([
                metadata_1.Directive({
                    selector: 'input[type=number][ng-control],input[type=number][ng-form-control],input[type=number][ng-model]',
                    host: {
                        '(change)': 'onChange($event.target.value)',
                        '(input)': 'onChange($event.target.value)',
                        '(blur)': 'onTouched()'
                    },
                    bindings: [NUMBER_VALUE_ACCESSOR]
                })
            ], NumberValueAccessor);
            NumberValueAccessor = NumberValueAccessor;
        }
    }
});
