// angular2/ts/src/core/forms/directives/shared.js
System.register(['angular2/src/core/facade/collection', 'angular2/src/core/facade/lang', 'angular2/src/core/facade/exceptions', '../validators', './default_value_accessor', './number_value_accessor', './checkbox_value_accessor', './select_control_value_accessor'], function(exports_1) {
    var collection_1, lang_1, exceptions_1, validators_1, default_value_accessor_1, number_value_accessor_1, checkbox_value_accessor_1, select_control_value_accessor_1;
    function controlPath(name, parent) {
        var p = collection_1.ListWrapper.clone(parent.path);
        p.push(name);
        return p;
    }
    exports_1("controlPath", controlPath);
    function setUpControl(control, dir) {
        if (lang_1.isBlank(control))
            _throwError(dir, "Cannot find control");
        if (lang_1.isBlank(dir.valueAccessor))
            _throwError(dir, "No value accessor for");
        control.validator = validators_1.Validators.compose([control.validator, dir.validator]);
        dir.valueAccessor.writeValue(control.value);
        // view -> model
        dir.valueAccessor.registerOnChange(newValue => {
            dir.viewToModelUpdate(newValue);
            control.updateValue(newValue, { emitModelToViewChange: false });
            control.markAsDirty();
        });
        // model -> view
        control.registerOnChange(newValue => dir.valueAccessor.writeValue(newValue));
        // touched
        dir.valueAccessor.registerOnTouched(() => control.markAsTouched());
    }
    exports_1("setUpControl", setUpControl);
    function _throwError(dir, message) {
        var path = dir.path.join(" -> ");
        throw new exceptions_1.BaseException(`${message} '${path}'`);
    }
    function setProperty(renderer, elementRef, propName, propValue) {
        renderer.setElementProperty(elementRef, propName, propValue);
    }
    exports_1("setProperty", setProperty);
    function isPropertyUpdated(changes, viewModel) {
        if (!collection_1.StringMapWrapper.contains(changes, "model"))
            return false;
        var change = changes["model"];
        if (change.isFirstChange())
            return true;
        return !lang_1.looseIdentical(viewModel, change.currentValue);
    }
    exports_1("isPropertyUpdated", isPropertyUpdated);
    // TODO: vsavkin remove it once https://github.com/angular/angular/issues/3011 is implemented
    function selectValueAccessor(dir, valueAccessors) {
        if (lang_1.isBlank(valueAccessors))
            return null;
        var defaultAccessor;
        var builtinAccessor;
        var customAccessor;
        valueAccessors.forEach(v => {
            if (v instanceof default_value_accessor_1.DefaultValueAccessor) {
                defaultAccessor = v;
            }
            else if (v instanceof checkbox_value_accessor_1.CheckboxControlValueAccessor || v instanceof number_value_accessor_1.NumberValueAccessor ||
                v instanceof select_control_value_accessor_1.SelectControlValueAccessor) {
                if (lang_1.isPresent(builtinAccessor))
                    _throwError(dir, "More than one built-in value accessor matches");
                builtinAccessor = v;
            }
            else {
                if (lang_1.isPresent(customAccessor))
                    _throwError(dir, "More than one custom value accessor matches");
                customAccessor = v;
            }
        });
        if (lang_1.isPresent(customAccessor))
            return customAccessor;
        if (lang_1.isPresent(builtinAccessor))
            return builtinAccessor;
        if (lang_1.isPresent(defaultAccessor))
            return defaultAccessor;
        _throwError(dir, "No valid value accessor for");
        return null;
    }
    exports_1("selectValueAccessor", selectValueAccessor);
    return {
        setters:[
            function (collection_1_1) {
                collection_1 = collection_1_1;
            },
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (exceptions_1_1) {
                exceptions_1 = exceptions_1_1;
            },
            function (validators_1_1) {
                validators_1 = validators_1_1;
            },
            function (default_value_accessor_1_1) {
                default_value_accessor_1 = default_value_accessor_1_1;
            },
            function (number_value_accessor_1_1) {
                number_value_accessor_1 = number_value_accessor_1_1;
            },
            function (checkbox_value_accessor_1_1) {
                checkbox_value_accessor_1 = checkbox_value_accessor_1_1;
            },
            function (select_control_value_accessor_1_1) {
                select_control_value_accessor_1 = select_control_value_accessor_1_1;
            }],
        execute: function() {
        }
    }
});
