// angular2/ts/src/animate/browser_details.js
System.register(['angular2/src/core/di', 'angular2/src/core/facade/math', 'angular2/src/core/dom/dom_adapter'], function(exports_1) {
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var di_1, math_1, dom_adapter_1;
    var BrowserDetails, RafQueue;
    return {
        setters:[
            function (di_1_1) {
                di_1 = di_1_1;
            },
            function (math_1_1) {
                math_1 = math_1_1;
            },
            function (dom_adapter_1_1) {
                dom_adapter_1 = dom_adapter_1_1;
            }],
        execute: function() {
            let BrowserDetails = class {
                constructor() {
                    this.elapsedTimeIncludesDelay = false;
                    this.doesElapsedTimeIncludesDelay();
                }
                /**
                 * Determines if `event.elapsedTime` includes transition delay in the current browser.  At this
                 * time, Chrome and Opera seem to be the only browsers that include this.
                 */
                doesElapsedTimeIncludesDelay() {
                    var div = dom_adapter_1.DOM.createElement('div');
                    dom_adapter_1.DOM.setAttribute(div, 'style', `position: absolute; top: -9999px; left: -9999px; width: 1px;
      height: 1px; transition: all 1ms linear 1ms;`);
                    // Firefox requires that we wait for 2 frames for some reason
                    this.raf(timestamp => {
                        dom_adapter_1.DOM.on(div, 'transitionend', (event) => {
                            var elapsed = math_1.Math.round(event.elapsedTime * 1000);
                            this.elapsedTimeIncludesDelay = elapsed == 2;
                            dom_adapter_1.DOM.remove(div);
                        });
                        dom_adapter_1.DOM.setStyle(div, 'width', '2px');
                    }, 2);
                }
                raf(callback, frames = 1) {
                    var queue = new RafQueue(callback, frames);
                    return () => queue.cancel();
                }
            };
            BrowserDetails = __decorate([
                di_1.Injectable()
            ], BrowserDetails);
            BrowserDetails = BrowserDetails;
            class RafQueue {
                constructor(callback, frames) {
                    this.callback = callback;
                    this.frames = frames;
                    this._raf();
                }
                _raf() {
                    this.currentFrameId = dom_adapter_1.DOM.requestAnimationFrame(timestamp => this._nextFrame(timestamp));
                }
                _nextFrame(timestamp) {
                    this.frames--;
                    if (this.frames > 0) {
                        this._raf();
                    }
                    else {
                        this.callback(timestamp);
                    }
                }
                cancel() {
                    dom_adapter_1.DOM.cancelAnimationFrame(this.currentFrameId);
                    this.currentFrameId = null;
                }
            }
        }
    }
});
