// angular2/ts/src/core/render/dom/events/key_events.js
System.register(['angular2/src/core/dom/dom_adapter', 'angular2/src/core/facade/lang', 'angular2/src/core/facade/collection', './event_manager', 'angular2/src/core/di'], function(exports_1) {
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var dom_adapter_1, lang_1, collection_1, event_manager_1, di_1;
    var modifierKeys, modifierKeyGetters, KeyEventsPlugin;
    return {
        setters:[
            function (dom_adapter_1_1) {
                dom_adapter_1 = dom_adapter_1_1;
            },
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (collection_1_1) {
                collection_1 = collection_1_1;
            },
            function (event_manager_1_1) {
                event_manager_1 = event_manager_1_1;
            },
            function (di_1_1) {
                di_1 = di_1_1;
            }],
        execute: function() {
            var modifierKeys = ['alt', 'control', 'meta', 'shift'];
            var modifierKeyGetters = {
                'alt': (event) => event.altKey,
                'control': (event) => event.ctrlKey,
                'meta': (event) => event.metaKey,
                'shift': (event) => event.shiftKey
            };
            let KeyEventsPlugin = class extends event_manager_1.EventManagerPlugin {
                constructor() {
                    super();
                }
                supports(eventName) {
                    return lang_1.isPresent(KeyEventsPlugin.parseEventName(eventName));
                }
                addEventListener(element, eventName, handler) {
                    var parsedEvent = KeyEventsPlugin.parseEventName(eventName);
                    var outsideHandler = KeyEventsPlugin.eventCallback(element, collection_1.StringMapWrapper.get(parsedEvent, 'fullKey'), handler, this.manager.getZone());
                    this.manager.getZone().runOutsideAngular(() => {
                        dom_adapter_1.DOM.on(element, collection_1.StringMapWrapper.get(parsedEvent, 'domEventName'), outsideHandler);
                    });
                }
                static parseEventName(eventName) {
                    var parts = eventName.toLowerCase().split('.');
                    var domEventName = parts.shift();
                    if ((parts.length === 0) ||
                        !(lang_1.StringWrapper.equals(domEventName, 'keydown') ||
                            lang_1.StringWrapper.equals(domEventName, 'keyup'))) {
                        return null;
                    }
                    var key = KeyEventsPlugin._normalizeKey(parts.pop());
                    var fullKey = '';
                    modifierKeys.forEach(modifierName => {
                        if (collection_1.ListWrapper.contains(parts, modifierName)) {
                            collection_1.ListWrapper.remove(parts, modifierName);
                            fullKey += modifierName + '.';
                        }
                    });
                    fullKey += key;
                    if (parts.length != 0 || key.length === 0) {
                        // returning null instead of throwing to let another plugin process the event
                        return null;
                    }
                    var result = collection_1.StringMapWrapper.create();
                    collection_1.StringMapWrapper.set(result, 'domEventName', domEventName);
                    collection_1.StringMapWrapper.set(result, 'fullKey', fullKey);
                    return result;
                }
                static getEventFullKey(event) {
                    var fullKey = '';
                    var key = dom_adapter_1.DOM.getEventKey(event);
                    key = key.toLowerCase();
                    if (lang_1.StringWrapper.equals(key, ' ')) {
                        key = 'space'; // for readability
                    }
                    else if (lang_1.StringWrapper.equals(key, '.')) {
                        key = 'dot'; // because '.' is used as a separator in event names
                    }
                    modifierKeys.forEach(modifierName => {
                        if (modifierName != key) {
                            var modifierGetter = collection_1.StringMapWrapper.get(modifierKeyGetters, modifierName);
                            if (modifierGetter(event)) {
                                fullKey += modifierName + '.';
                            }
                        }
                    });
                    fullKey += key;
                    return fullKey;
                }
                static eventCallback(element, fullKey, handler, zone) {
                    return (event) => {
                        if (lang_1.StringWrapper.equals(KeyEventsPlugin.getEventFullKey(event), fullKey)) {
                            zone.run(() => handler(event));
                        }
                    };
                }
                static _normalizeKey(keyName) {
                    // TODO: switch to a StringMap if the mapping grows too much
                    switch (keyName) {
                        case 'esc':
                            return 'escape';
                        default:
                            return keyName;
                    }
                }
            };
            KeyEventsPlugin = __decorate([
                di_1.Injectable()
            ], KeyEventsPlugin);
            KeyEventsPlugin = KeyEventsPlugin;
        }
    }
});
