// angular2/ts/src/core/compiler/source_module.js
System.register(['angular2/src/core/facade/lang'], function(exports_1) {
    var lang_1;
    var MODULE_REGEXP, SourceModule, SourceExpression, SourceExpressions, SourceWithImports;
    function moduleRef(moduleUrl) {
        return `#MODULE[${moduleUrl}]`;
    }
    exports_1("moduleRef", moduleRef);
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            }],
        execute: function() {
            var MODULE_REGEXP = /#MODULE\[([^\]]*)\]/g;
            class SourceModule {
                constructor(moduleUrl, sourceWithModuleRefs) {
                    this.moduleUrl = moduleUrl;
                    this.sourceWithModuleRefs = sourceWithModuleRefs;
                }
                getSourceWithImports() {
                    var moduleAliases = {};
                    var imports = [];
                    var newSource = lang_1.StringWrapper.replaceAllMapped(this.sourceWithModuleRefs, MODULE_REGEXP, (match) => {
                        var moduleUrl = match[1];
                        var alias = moduleAliases[moduleUrl];
                        if (lang_1.isBlank(alias)) {
                            if (moduleUrl == this.moduleUrl) {
                                alias = '';
                            }
                            else {
                                alias = `import${imports.length}`;
                                imports.push([moduleUrl, alias]);
                            }
                            moduleAliases[moduleUrl] = alias;
                        }
                        return alias.length > 0 ? `${alias}.` : '';
                    });
                    return new SourceWithImports(newSource, imports);
                }
            }
            SourceModule = SourceModule;
            class SourceExpression {
                constructor(declarations, expression) {
                    this.declarations = declarations;
                    this.expression = expression;
                }
            }
            SourceExpression = SourceExpression;
            class SourceExpressions {
                constructor(declarations, expressions) {
                    this.declarations = declarations;
                    this.expressions = expressions;
                }
            }
            SourceExpressions = SourceExpressions;
            class SourceWithImports {
                constructor(source, imports) {
                    this.source = source;
                    this.imports = imports;
                }
            }
            SourceWithImports = SourceWithImports;
        }
    }
});
