// angular2/ts/src/core/pipes/async_pipe.js
System.register(['angular2/src/core/facade/lang', 'angular2/src/core/facade/async', 'angular2/src/core/metadata', 'angular2/src/core/di', 'angular2/src/core/change_detection', './invalid_pipe_argument_exception'], function(exports_1) {
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var lang_1, async_1, metadata_1, di_1, change_detection_1, invalid_pipe_argument_exception_1;
    var ObservableStrategy, PromiseStrategy, _promiseStrategy, _observableStrategy, AsyncPipe;
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            },
            function (async_1_1) {
                async_1 = async_1_1;
            },
            function (metadata_1_1) {
                metadata_1 = metadata_1_1;
            },
            function (di_1_1) {
                di_1 = di_1_1;
            },
            function (change_detection_1_1) {
                change_detection_1 = change_detection_1_1;
            },
            function (invalid_pipe_argument_exception_1_1) {
                invalid_pipe_argument_exception_1 = invalid_pipe_argument_exception_1_1;
            }],
        execute: function() {
            class ObservableStrategy {
                createSubscription(async, updateLatestValue) {
                    return async_1.ObservableWrapper.subscribe(async, updateLatestValue, e => { throw e; });
                }
                dispose(subscription) { async_1.ObservableWrapper.dispose(subscription); }
                onDestroy(subscription) { async_1.ObservableWrapper.dispose(subscription); }
            }
            class PromiseStrategy {
                createSubscription(async, updateLatestValue) {
                    return async.then(updateLatestValue);
                }
                dispose(subscription) { }
                onDestroy(subscription) { }
            }
            var _promiseStrategy = new PromiseStrategy();
            var _observableStrategy = new ObservableStrategy();
            /**
             * The `async` pipe subscribes to an Observable or Promise and returns the latest value it has
             * emitted.
             * When a new value is emitted, the `async` pipe marks the component to be checked for changes.
             *
             * # Example
             * The example below binds the `time` Observable to the view. Every 500ms, the `time` Observable
             * updates the view with the current time.
             *
             * ```
             * import {Observable} from 'angular2/core';
             * @Component({
             *   selector: "task-cmp",
             *   template: "Time: {{ time | async }}"
             * })
             * class Task {
             *   time = new Observable<number>(observer => {
             *     setInterval(_ =>
             *       observer.next(new Date().getTime()), 500);
             *   });
             * }
             * ```
             */
            let AsyncPipe = class {
                constructor(_ref) {
                    /** @internal */
                    this._latestValue = null;
                    /** @internal */
                    this._latestReturnedValue = null;
                    /** @internal */
                    this._subscription = null;
                    /** @internal */
                    this._obj = null;
                    this._strategy = null;
                    this._ref = _ref;
                }
                onDestroy() {
                    if (lang_1.isPresent(this._subscription)) {
                        this._dispose();
                    }
                }
                transform(obj, args) {
                    if (lang_1.isBlank(this._obj)) {
                        if (lang_1.isPresent(obj)) {
                            this._subscribe(obj);
                        }
                        return null;
                    }
                    if (obj !== this._obj) {
                        this._dispose();
                        return this.transform(obj);
                    }
                    if (this._latestValue === this._latestReturnedValue) {
                        return this._latestReturnedValue;
                    }
                    else {
                        this._latestReturnedValue = this._latestValue;
                        return change_detection_1.WrappedValue.wrap(this._latestValue);
                    }
                }
                /** @internal */
                _subscribe(obj) {
                    this._obj = obj;
                    this._strategy = this._selectStrategy(obj);
                    this._subscription =
                        this._strategy.createSubscription(obj, value => this._updateLatestValue(obj, value));
                }
                /** @internal */
                _selectStrategy(obj) {
                    if (lang_1.isPromise(obj)) {
                        return _promiseStrategy;
                    }
                    else if (async_1.ObservableWrapper.isObservable(obj)) {
                        return _observableStrategy;
                    }
                    else {
                        throw new invalid_pipe_argument_exception_1.InvalidPipeArgumentException(AsyncPipe, obj);
                    }
                }
                /** @internal */
                _dispose() {
                    this._strategy.dispose(this._subscription);
                    this._latestValue = null;
                    this._latestReturnedValue = null;
                    this._subscription = null;
                    this._obj = null;
                }
                /** @internal */
                _updateLatestValue(async, value) {
                    if (async === this._obj) {
                        this._latestValue = value;
                        this._ref.markForCheck();
                    }
                }
            };
            AsyncPipe = __decorate([
                metadata_1.Pipe({ name: 'async', pure: false }),
                di_1.Injectable()
            ], AsyncPipe);
            AsyncPipe = AsyncPipe;
        }
    }
});
